src/org/ojalgo/scalar/RationalNumber.java

```diff
@@ -177,50 +174,42 @@ public static RationalNumber valueOf(final double value) {
-        //        final long significand = (exponent == 0 ? (valBits & ((1L << 52) - 1)) << 1 : (valBits & ((1L << 52) - 1)) | (1L << 52));
-        //        exponent -= 1075;
-        //
-        //        if (exponent < 0) {
-        //            return RationalNumber.of(sign * significand, 1L << exponent);
-        //        } else {
-        //            return RationalNumber.of(sign * significand * (1L << exponent), 1L);
-        //        }
-
-        // Code below doesn't work for 0 and NaN - just check before
-
-        //        final long bits = Double.doubleToLongBits(value);
-        //
-        //        final long sign = bits >>> 63;
-        //        final long exponent = ((bits >>> 52) ^ (sign << 11)) - 1023;
-        //        final long fraction = bits << 12; // bits are "reversed" but that's not a problem
-        //
-        //        long a = 1L;
-        //        long b = 1L;
-        //
-        //        for (int i = 63; i >= 12; i--) {
-        //            a = (a * 2) + ((fraction >>> i) & 1);
-        //            b *= 2;
-        //        }
-        //
-        //        if (exponent > 0) {
-        //            a *= 1 << exponent;
-        //        } else {
-        //            b *= 1 << -exponent;
-        //        }
-        //
-        //        if (sign == 1) {
-        //            a *= -1;
-        //        }
-        //
-        //        // Here you have to simplify the fraction
-        //
-        //        System.out.println(a + "/" + b);
-
-        return RationalNumber.valueOf(BigDecimal.valueOf(value));
+        final long bits = Double.doubleToLongBits(value);
+
+        // Please refer to {@link Double#doubleToLongBits(long)} javadoc
+        final int s = ((bits >> 63) == 0) ? 1 : -1;
+        final int e = (int) ((bits >> 52) & 0x7ffL);
+        long m = (e == 0) ?
+                (bits & 0xfffffffffffffL) << 1 :
+                (bits & 0xfffffffffffffL) | 0x10000000000000L;
+        // Now we're looking for s * m * 2^{e - 1075}, 1075 being bias of 1023 plus 52 positions of binary fraction
+
+        int exponent = e - 1075;
+
+        if (exponent >= 0) {
+            long numerator = m << exponent;
+            if (numerator >> exponent != m) {
+                return s > 0 ? RationalNumber.POSITIVE_INFINITY : RationalNumber.NEGATIVE_INFINITY;
+            }
+            return new RationalNumber(s * numerator, 1L);
+        }
+
+        // Since denominator is a power of 2, GCD can only be power of two, so we simplify by dividing by 2 repeatedly
+        while ((m & 1) == 0 && exponent < 0) {
+            m >>= 1;
+            exponent++;
+        }
+        // Avoiding the the denominator overflow
+        if (-exponent >= MAX_BITS) {
+            BigInteger denom = BigInteger.ONE.shiftLeft(-exponent);
+            BigInteger maxlong = BigInteger.valueOf(Long.MAX_VALUE);
+            BigInteger factor = denom.divide(maxlong);
```

Comment:
Now instead of dividing by `Long.MAX_VALUE`, `2^63 - 1`, we could divide by `2^63`: off-by-1 error here is not likely to matter, and the operation becomes enormously cheaper. Still, that sounds like a kludge.

A better approach would be to use rational fraction aproximation, as in https://uk.mathworks.com/help/matlab/ref/rat.html?s_tid=gn_loc_drop - after all, we're not that bound to powers of 2 and 10. 

Alternatively, we can keep the binary exponent around. That would let us keep the whole accuracy of `double` and then some.